compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Upgrade.hs

```diff
@@ -0,0 +1,254 @@
+    -> m (HMS.HashMap k (Upgrading a), HMS.HashMap k a)
+checkDeleted handle upgrade = do
+    let (deleted, existing, new) = extractDelExistNew upgrade
+    case HMS.toList deleted of
+      ((_, head):_) -> throwWithContext $ handle head
+      _ -> pure (existing, new)
+
+checkModule :: MonadGamma m => Upgrading LF.Module -> m ()
+checkModule module_ = do
+    (existingTemplates, _new) <- checkDeleted (EUpgradeMissingTemplate . NM.name) $ NM.toHashMap . moduleTemplates <$> module_
+    forM_ existingTemplates $ \template ->
+        withContext
+            (ContextTemplate (present module_) (present template) TPWhole)
+            (checkTemplate module_ template)
+
+    -- TODO: Handle deleted datatypes (not belonging to templates or choices)
+    let (_dtDel, dtExisting, _dtNew) = extractDelExistNew $ NM.toHashMap . moduleDataTypes <$> module_
+
+    -- For a datatype, derive its context
+    let deriveChoiceInfo :: LF.Module -> HMS.HashMap LF.TypeConName (LF.Template, LF.TemplateChoice)
+        deriveChoiceInfo module_ = HMS.fromList $ do
+            template <- NM.toList (moduleTemplates module_)
+            choice <- NM.toList (tplChoices template)
+            TCon dtName <- [snd (chcArgBinder choice)] -- Choice inputs should always be a list
+            pure (qualObject dtName, (template, choice))
+        allChoiceReturnTCons :: LF.Module -> HMS.HashMap LF.TypeConName (LF.Template, LF.TemplateChoice)
+        allChoiceReturnTCons module_ = HMS.fromList $ do
+            template <- NM.toList (moduleTemplates module_)
+            choice <- NM.toList (tplChoices template)
+            dtName <- flip cata (chcReturnType choice) $ \case
+                TConF dtName -> [dtName]
+                rest -> fold rest
+            pure (qualObject dtName, (template, choice))
+        dataTypeOrigin
+            :: DefDataType -> Module
+            -> (UpgradedRecordOrigin, Context)
+        dataTypeOrigin dt module_
+            | Just template <- NM.name dt `NM.lookup` moduleTemplates module_ =
+                ( TemplateBody (NM.name dt)
+                , ContextTemplate module_ template TPWhole
+                )
+            | Just (template, choice) <- NM.name dt `HMS.lookup` deriveChoiceInfo module_ =
+                ( TemplateChoiceInput (NM.name template) (NM.name choice)
+                , ContextTemplate module_ template (TPChoice choice)
+                )
+            | Just (template, choice) <- NM.name dt `HMS.lookup` allChoiceReturnTCons module_ =
+                ( TemplateChoiceOutput (NM.name template) (NM.name choice) (NM.name dt)
+                , ContextTemplate module_ template (TPChoice choice)
+                )
+            | otherwise = (TopLevel, ContextDefDataType module_ dt)
+
+    forM_ dtExisting $ \dt ->
+        -- Get origin/context for each datatype in both past and present
+        let origin = dataTypeOrigin <$> dt <*> module_
+        in
+        -- If origins don't match, record has changed origin
```

Comment:
I tried for a bit to make this work because I like the idea of telling a user where the datatype shows up, but in the end it didn't work out so I've scrapped the `TemplateChoiceOutput` variant and replaced it with "shouldCheck"